src/com/facebook/buck/intellij/ideabuck/src/com/facebook/buck/intellij/ideabuck/actions/select/RunSelectedTestAction.java

```diff
@@ -0,0 +1,172 @@
+    PsiMethod method = null;
+    PsiClass psiClass = null;
+    if (element instanceof PsiMethod) {
+      method = (PsiMethod) element;
+      psiClass = method.getContainingClass();
+    }
+    if (element instanceof PsiClass) {
+      psiClass = (PsiClass) element;
+    }
+    final Project project = event.getProject();
+    if (project == null) {
+      return;
+    }
+    Editor editor = FileEditorManager.getInstance(project).getSelectedTextEditor();
+    if (editor == null) {
+      return;
+    }
+    final Document document = editor.getDocument();
+    VirtualFile virtualFile = FileDocumentManager.getInstance().getFile(document);
+
+    VirtualFile buckFile = BuckFileUtil.getBuckFile(virtualFile);
+    if (buckFile != null) {
+      String name;
+      Optional<String> testSelectors = Optional.empty();
+      if (method != null) {
+        name = psiClass.getName() + "#" + method.getName();
+        testSelectors = Optional.of(psiClass.getName() + "#" + method.getName());
+      } else {
+        name = element.getContainingFile().getName();
+        if (element instanceof PsiClass) {
+          testSelectors = Optional.of(psiClass.getQualifiedName());
+        }
+      }
+      createTestConfigurationFromContext(
+          name,
+          testSelectors,
+          project,
+          virtualFile,
+          buckFile,
+          debug);
+    }
+  }
+
+  /**
+   * Create a test Configuration and run or debug it in the ide.
+   *
+   * @param name a {@link String} representing the name of the configuration.
+   * @param testSelectors a {@link String} representing optional testSelectors for filtering.
+   * @param project {@link Project} then intellij project corresponding the the file or module under test.
+   * @param containingFile {@link VirtualFile} the file that containing the impacted tests.
+   * @param buckFile {@link VirtualFile} the file representing the buck File.
+   * @param debug a {@link boolean} used to signify run or debug.
+   */
+  private void createTestConfigurationFromContext(
+      @Nonnull String name,
+      final Optional<String> testSelectors,
+      final Project project,
+      VirtualFile containingFile,
+      VirtualFile buckFile,
+      boolean debug) {
+    BuckBuildManager buildManager = BuckBuildManager.getInstance(project);
+    BuckQueryCommandHandler handler =
+      new BuckQueryCommandHandler(
+          project,
+          buckFile.getParent(),
+          BuckCommand.QUERY, new Function<List<String>, Void>() {
+        @Nullable
+        @Override
+        public Void apply(@Nullable List<String> strings) {
+          RunManagerImpl runManager = (RunManagerImpl) RunManager.getInstance(project);
+          TestConfigurationType type = new TestConfigurationType();
+          RunnerAndConfigurationSettingsImpl runnerAndConfigurationSettings
+              = (RunnerAndConfigurationSettingsImpl)runManager.createRunConfiguration(name, type.getConfigurationFactories()[0]);
```

Comment:
What would happen if `type.getConfigurationFactories()` is empty? And if first element is not `RunnerAndConfigurationSettingsImpl`?